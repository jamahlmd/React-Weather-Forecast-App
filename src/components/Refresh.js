import React, {Component} from 'react';
import { Progress } from 'antd';
import propTypes from 'prop-types';


class Refresh extends Component {
    state = {
        running: this.props.refresh,
        elapsedTime: 0,
        previousTime: 0

    };



    //Start the ticking
    componentDidMount = () => {
        this.onStart();
        this.interval = setInterval(this.onTick, 100);
    };



    //If the refresh prop changed, save it to state so that this component know whether to tick or not
    //If the lastSearch prop changed (means the user entered a new city)
    //reset the timer to keep it in sync with the fetchdata interval
    componentDidUpdate = (prevProps, prevState) => {
        prevProps.refresh !== this.props.refresh && this.setState({running: this.props.refresh});

        prevProps.lastSearch !== this.props.lastSearch && this.onReset();
    };

    componentWillUnmount = () =>
        clearInterval(this.interval);

    //If the this.state.running is true, keep ticking until the timer hits 30 seconds and then reset it
    //Else reset the timer
    onTick = () => {
        if(this.state.running){
            const now = Date.now();

            if(Math.floor(this.state.elapsedTime / 1000) >= 30) {
                this.onReset();
            } else {
                this.setState({
                    previousTime: now,
                    //Added half a second because I found that the tickinginterval didn't sync really well with the fetchdatainterval
                    //To sync them better I think I had to keep this state in App.js
                    elapsedTime: this.state.elapsedTime + (now - this.state.previousTime) + 0.45
                })
            }
        } else {
            this.onReset();
        }
    };

    onStart = () => {
        this.setState({
            running: true,
            previousTime: Date.now(),
        })
    };

    onReset = () => {
        this.setState({
            elapsedTime: 0,
            previousTime: Date.now()
        })
    };


    render(){


        let seconds = 30 - Math.floor(this.state.elapsedTime / 1000);

        return(
            <div className="responsive_progressbar" style={{width: 270, paddingTop: '3px'}}>
                <span>Refresh in: {seconds}</span>
                <Progress
                    percent={100 - (seconds * 3.34)}
                    size="small"
                    format={ () => null}
                />
            </div>
        )
    }
}

Refresh.propTypes = {
  refresh: propTypes.bool.isRequired,
  lastSearch: propTypes.string.isRequired
};

export default Refresh;