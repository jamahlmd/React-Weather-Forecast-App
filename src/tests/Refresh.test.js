import React from 'react';
import Refresh from '../components/Refresh';

const wrapper = shallow(<Refresh refresh={true} lastSearch={'London'}/>);


test('Should render Refresh correctly', () => {
    expect(toJson(wrapper)).toMatchSnapshot();
});


test('Should set state properly according to props', () => {
   expect(wrapper.state('running')).toBe(true);
});


test('Should update state properly on componentDidUpdate', () => {
    wrapper.setProps({refresh: false});
    expect(wrapper.state('running')).toBe(false);
});



test('Should refresh timer when city changes', () => {
    wrapper.setProps({lastSearch: 'Prague'});
    expect(wrapper.state('elapsedTime')).toBe(0);
});