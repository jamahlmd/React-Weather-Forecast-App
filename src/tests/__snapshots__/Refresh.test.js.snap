// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render Refresh correctly 1`] = `
<div
  style={
    Object {
      "paddingTop": "3px",
      "width": 270,
    }
  }
>
  Refresh in: 
  30
  <Progress
    format={[Function]}
    percent={9.999999974752427e-7}
    prefixCls="ant-progress"
    showInfo={true}
    size="small"
    trailColor="#f3f3f3"
    type="line"
  />
</div>
`;
